providers/kind/kind.sh

```diff
@@ -0,0 +1,173 @@
+		echo "Waiting for kind to be ready ..."
+		sleep 10
+	done
+	echo "Waiting for dns to be ready ..."
+	kubectl wait -n kube-system --timeout=12m --for=condition=Ready -l k8s-app=kube-dns pods
+}
+#
+_prepare_config() {
+	local default_registry_name="kind-registry"
+	local default_registry_port="5001"
+
+	info "Building manifests..."
+	cp $KIND_MANIFESTS_DIR/kind.yml "$KIND_CONFIG_YAML"
+
+	sed "s/$default_registry_name/${KIND_REGISTRY_NAME}/g" "$KIND_CONFIG_YAML" >"${KIND_CONFIG_YAML}.tmp" &&
+		mv "${KIND_CONFIG_YAML}.tmp" "${KIND_CONFIG_YAML}"
+
+	sed "s/$default_registry_port/${REGISTRY_PORT}/g" "$KIND_CONFIG_YAML" >"${KIND_CONFIG_YAML}.tmp" &&
+		mv "${KIND_CONFIG_YAML}.tmp" "$KIND_CONFIG_YAML"
+
+	cp "$KIND_MANIFESTS_DIR/local-registry.yml" "$KIND_REGISTRY_YAML"
+
+	sed "s/$default_registry_name/${KIND_REGISTRY_NAME}/g" "${KIND_REGISTRY_YAML}" >"${KIND_REGISTRY_YAML}.tmp" &&
+		mv "${KIND_REGISTRY_YAML}.tmp" "${KIND_REGISTRY_YAML}"
+
+	sed "s/$default_registry_port/${REGISTRY_PORT}/g" "$KIND_REGISTRY_YAML" >"${KIND_REGISTRY_YAML}.tmp" &&
+		mv "${KIND_REGISTRY_YAML}.tmp" "$KIND_REGISTRY_YAML"
+}
+
+_setup_kind() {
+	info "Starting kind with cluster name \"${KIND_CLUSTER_NAME}\""
+
+	$KIND create cluster --name="${KIND_CLUSTER_NAME}" -v6 --config="$KIND_CONFIG_YAML"
+	$KIND get kubeconfig --name="${KIND_CLUSTER_NAME}" >${KIND_DIR}/.kubeconfig
+
+	kind_wait_up
+	# wait until k8s pods are running
+}
+
+_run_kind_registry() {
+	until [ -z "$($CTR_CMD ps -a | grep "${KIND_REGISTRY_NAME}")" ]; do
+		$CTR_CMD stop "${KIND_REGISTRY_NAME}" || true
+		$CTR_CMD rm "${KIND_REGISTRY_NAME}" || true
+		sleep 5
+	done
+
+	$CTR_CMD run \
+		-d --restart=always \
+		-p "127.0.0.1:${REGISTRY_PORT}:5000" \
+		--name "${KIND_REGISTRY_NAME}" \
+		registry:2
+	# connect the registry to the cluster network if not already connected
+	$CTR_CMD network connect "${KIND_DEFAULT_NETWORK}" "${KIND_REGISTRY_NAME}" || true
+	kubectl apply -f "${KIND_DIR}"/local-registry.yml
+}
+
+### exported functions
+
+kind_print_config() {
+	cat <<-EOF
+		KIND
+		──────────────────────────────────────
+		Binary:  $KIND
+		Version: $KIND_VERSION
+
+		Cluster Name : $KIND_CLUSTER_NAME
+		Config file  : $KIND_CONFIG_YAML
+	EOF
+
+}
+
+kind_up() {
+	mkdir -p "$KIND_DIR"
+
+	kind_install
+	_prepare_config
+	_setup_kind
+	wait_for_cluster_ready
+	_run_kind_registry
+}
+
+kind_down() {
+	kind_install
+	if ! $KIND get clusters 2>/dev/null | grep -q "${KIND_CLUSTER_NAME}"; then
+		info "No kind cluster $KIND_CLUSTER_NAME found; skipping deletion"
+		return
+	fi
+
+	# NOTE: Avoid failing an entire test run just because of a deletion error
+	$KIND delete cluster --name="${KIND_CLUSTER_NAME}" || true
+
+	$CTR_CMD rm -f "${KIND_REGISTRY_NAME}" >/dev/null
+	rm -f "$KIND_CONFIG_YAML" "$KIND_REGISTRY_YAML"
+	find "${KIND_DIR}" -maxdepth 1 -name '.*' -delete
```

Comment:
I think we can delete the complete `tmp` directory instead of removing files from `tmp/kind` as it won't make sense to keep it

Reply:
I am against deleting the entirety of tmp dir since (folks, aka me) may keep other files there :p .. also it is good idea to preserve tmp/bin so that the binaries need not be downloaded against.